include/seqan3/alignment/matrix/detail/coordinate_matrix.hpp

```diff
@@ -12,43 +12,172 @@
+ * to form an indexed alignment matrix.
+ *
+ * This matrix is cheap as it stores only the dimensions of the matrix and does not allocate any memory for the
+ * coordinates. It uses the seqan3::detail::matrix_coordinate_iterator to iterate
+ * over the columns of this virtual matrix. When the seqan3::detail::coordinate_matrix::matrix_coordinate_iterator
+ * is dereferenced it returns an on-the-fly constructed range
+ * ([prvalue](https://en.cppreference.com/w/cpp/language/value_category))
+ * representing the current indexed column. The reference type of this range is seqan3::detail::matrix_coordinate.
+ *
+ * ### Simd mode
+ *
+ * If the `index_t` is a simd vector the matrix implements a vectorised coordinate matrix instead. In this case the
+ * matrix uses a seqan3::detail::coordinate_matrix::matrix_coordinate_simd_iterator to iterate over the vectorised
+ * matrix. When the seqan3::detail::coordinate_matrix::matrix_coordinate_iterator is dereferenced it returns an
+ * on-the-fly constructed range ([prvalue](https://en.cppreference.com/w/cpp/language/value_category)) representing the
+ * current simd coordinate column. The reference type of this range is seqan3::detail::simd_matrix_coordinate.
  */
-struct convert_to_matrix_coordinate
+template <typename index_t>
+//!\cond
+    requires std::integral<index_t> ||
+             (simd::simd_concept<index_t> && requires (simd::simd_traits<index_t> traits)
+             {
+                 requires std::integral<typename decltype(traits)::scalar_type>;
+             })
+//!\endcond
+class coordinate_matrix
 {
-    //!\brief The index of the represented column.
-    size_t column_index{0};
+private:
+    /*!\brief A function object that converts a column index and a row index to a seqan3::detail::matrix_coordinate.
+     * \details
+     *
+     * This helper function object is used for the seqan3::detail::coordinate_matrix::coordinate_matrix_iterator and
+     * seqan3::detail::coordinate_matrix::coordinate_matrix_simd_iterator to convert the column and row index into a
+     * seqan3::detail::matrix_index.
+     */
+    struct convert_to_matrix_coordinate
+    {
+        //!\brief The index of the represented column.
+        index_t column_index{};
+
+        /*!\brief The conversion operator
+         *
+         * \param[in] row_index The index of the represented row.
+         *
+         * \returns seqan3::detail::matrix_index for the current column and row index.
+         */
+        auto operator()(index_t const row_index) noexcept
+        {
+            return matrix_index{row_index_type{row_index}, column_index_type{column_index}};
+        }
+    };
+
+    // The coordinate matrix iterator for simd index types.
+    template <simd_concept simd_index_t>
+    class coordinate_matrix_simd_iterator;
+    // The coordinate matrix iterator for regular integral types.
+    class coordinate_matrix_iterator;
```

Comment:
Why not have only one class? They are nearly identical!

Reply:
I am in general not a fan of to much if constexpr. It always means that such a class maintains two different states and thus will always lead to some errors in the future. But I will change this since this will be an endless discussion.